### ISSUE TITLE
Problem with DNS servers

### ISSUE THREAD

**Issue by gilza888** [amnezia-vpn/amnezia-client#708]
I can’t access websites when VPN is enabled.
Pings IP addresses, but when trying to ping google.com or other domains it gives an error. VPN works on other devices

Amnezia: 4.4.0.0
Linux: KaOS rolling

**Comment by Guest007** [comment]
4.5.0.0 from github
Manjaro Linux
AmneziaWG - DNS not working. AmneziaDNS nor 8.8.8.8/1.1.1.1 etc
Amnezia OpenVPN over Cloak works well

**Comment by pokamest** [comment]
Please test https://github.com/amnezia-vpn/amnezia-client/releases/tag/4.5.3.0

**Comment by Guest007** [comment]
> Please test https://github.com/amnezia-vpn/amnezia-client/releases/tag/4.5.3.0

Thank you. But it's not work for me now

Connect established, but dns not acceptable.

**Comment by DmitriyMolch** [comment]
Not working for me 
![image](https://github.com/amnezia-vpn/amnezia-client/assets/10967748/c57da621-47c5-4334-a890-d6daa1fb4a9b)
![image](https://github.com/amnezia-vpn/amnezia-client/assets/10967748/5f4d942b-67f6-4c02-beb5-a23e08d227f2)

**Comment by SmirnovKirill** [comment]
You can make a backup file, find sections with dns (nDNS = ...) and \"dns1\": \"\",\n        \"dns2\": \"\" and change them to any dns you like and then import this backup file. This worked for me.

**Comment by evgeny-gerasimov** [comment]
This problem also remains in version 4.6.0.3 (AmneziaWG).
OS: Manjaro linux 24.0.5.
![01_08_2024_15-10-42](https://github.com/user-attachments/assets/21c854bc-7a4f-43d9-a200-490fe5eabf53)
![01_08_2024_15-10-26](https://github.com/user-attachments/assets/e33004e4-b013-4b7a-aeeb-344df5466618)

**Comment by GILLIBON** [comment]
Same problem with Arch Linux.
AmneziaVPN version 4.6.0.3.
Unable to resolve domain names with OpenVPN and AmneziaWG. OpenVPN over Cloak works fine.

**Comment by teamfighter** [comment]
Same problem here.

**Comment by FeorPrime** [comment]
Linux 6.1.0-22-amd64 Debian 6.1.94-1 (2024-06-21) x86_64 GNU/Linux

both on 4.6.0.3 and 4.5.0.0 has the same problem. 

Vpn on:
```sh
❯ nslookup google.com
;; communications error to 192.168.50.1#53: connection refused
;; communications error to 192.168.50.1#53: connection refused
;; communications error to 192.168.50.1#53: connection refused
;; communications error to 8.8.8.8#53: connection refused
;; communications error to 8.8.4.4#53: connection refused
;; no servers could be reached
```
and whithout vpn:
```sh
❯ nslookup google.com
Server:		192.168.50.1
Address:	192.168.50.1#53

Non-authoritative answer:
Name:	google.com
Address: 142.250.186.142
Name:	google.com
Address: 2a00:1450:4026:803::200e
```
i decided to check:
```sh
vpn off:

❯ nc -vz 142.250.186.142 443
Connection to 142.250.186.142 443 port [tcp/https] succeeded!
❯ nc -vz 142.250.186.142 80
Connection to 142.250.186.142 80 port [tcp/http] succeeded!
❯ nc -vz 8.8.8.8 53
Connection to 8.8.8.8 53 port [tcp/domain] succeeded!

vpn on:

❯ nc -vz 8.8.8.8 53
nc: connect to 8.8.8.8 port 53 (tcp) failed: Connection refused
❯ nc -vz 142.250.186.142 80
Connection to 142.250.186.142 80 port [tcp/http] succeeded!
❯ nc -vz 142.250.186.142 443
Connection to 142.250.186.142 443 port [tcp/https] succeeded!
```

### TL;DR : so, my workaround was this:

```sh
❯ sudo iptables -L | grep DNS
Chain amnvpn.310.blockDNS (1 references)
Chain amnvpn.320.allowDNS (1 references)
Chain amnvpn.a.310.blockDNS (1 references)
amnvpn.310.blockDNS  all  --  anywhere             anywhere            
Chain amnvpn.a.320.allowDNS (1 references)
amnvpn.320.allowDNS  all  --  anywhere             anywhere            
amnvpn.a.320.allowDNS  all  --  anywhere             anywhere            
amnvpn.a.310.blockDNS  all  --  anywhere             anywhere 

```
I thought that it looks wierd, and i decide to delete some rules 1 by 1 (cuz i have no idea how iptables works):

```sh
sudo iptables -F amnvpn.a.310.blockDNS
```

and now everything is working, except amnezia dns:

```sh
❯ nc -vz 8.8.8.8 53
Connection to 8.8.8.8 53 port [tcp/domain] succeeded!
❯ nslookup google.com
Server:		192.168.50.1
Address:	192.168.50.1#53

Non-authoritative answer:
Name:	google.com
Address: 216.58.209.206
Name:	google.com
Address: 2a00:1450:4026:803::200e
```

### Still need to hit `sudo iptables -F amnvpn.a.310.blockDNS` each time i need to use vpn!!

**Comment by f0ma** [comment]
If I understand the logic in `client/platforms/linux/daemon/dnsutilslinux.cpp` correctly `amnezia-client` tries to send information about new DNS servers via DBUS to local resolver (`systemd-resolved`), and blocks all other DNS requests with the firewall rule `amnvpn.a.310.blockDNS`. If local resolver is disabled (default on Arch and Manjaro), it causes an error in DNS name resolution.

You can start `systemd-resolved` before connecting to VPN to fix this issue:
`sudo systemctl start systemd-resolved`

or enable it on system boot
`sudo systemctl enable systemd-resolved`

**Comment by ghost** [comment]
It didn't work for me with Kali Linux, I installed systemd-resolved and started service, it's running, but when I turn on VPN and try to find IP of google.com it writes:
nslookup google.com                   
Server:         127.0.0.53
Address:        127.0.0.53#53

** server can't find google.com: REFUSED

Moreover, after VPN turned off, DNS still was not working. I had to write nameservers to resolv.conf file manually to restore Internet without VPN.

**Comment by ghost** [comment]
I solved issue by changing nameserver 127.0.0.53 to nameserver 172.29.172.254 (it's the address of Amnezia DNS) in /etc/resolv/conf and **not** reloading systemd-resolved after making changes. It seems that there is no DNS leak and everything works fine.

**Comment by Guest007** [comment]
With 4.7.0.0 AmneziaWG started to work
OS: Manjaro Linux
Client app from Releases page here

**Comment by Kolbasyatin** [comment]
Doesn't work for me when the systemd-resolved and the VPN are active both.
I have to do `systemctl disable systemd-resolved` `systemctl stop systemd-resolved` and just then DNS resolve works again. 
EndeavourOS
Amnezia 4.7.0.0

**Comment by odduck41** [comment]
The same problem...

**Comment by blobski** [comment]
`sudo apt install openvpn-systemd-resolved` did it for me (Ubuntu 24).
Don't know what is DNS leak though :)

**Comment by missingnx** [comment]
> I solved issue by changing nameserver 127.0.0.53 to nameserver 172.29.172.254 (it's the address of Amnezia DNS) in /etc/resolv/conf and **not** reloading systemd-resolved after making changes. It seems that there is no DNS leak and everything works fine.

@dsonozaki
Solve this on Kali by using 8.8.8.8 as nameserver in `/etc/resolv.conf`
It works w/ & w/o Amnezia

P.S.: May be this is insecure, idk. Sry, i'm newbie

**Comment by skydiver89** [comment]
Had same issue with XRay protocol (OpenVPN over Cloak worked fine) Amnezia 4.8.2.3, Amnezia 4.7.0.0 on Ubuntu 24.04.
Solution which helped for me:  click button "Save" in "XRay server settings".

**Comment by msgtv** [comment]
I installed `systemd-networkd` and created a soft link to `/etc/resolv.conf` and it worked on Alt Linux Workstation 10
```
sudo apt-get install systemd-networkd
sudo systemctl enable systemd-resolved
sudo systemctl start systemd-resolved
sudo ln -sf /run/systemd/resolve/stub-resolv.conf /etc/resolv.conf
```

**Comment by seleniumforest** [comment]
kubuntu 24.04, fresh install, this helped me
```
sudo systemctl stop systemd-resolved
sudo systemctl disable systemd-resolved
sudo apt remove systemd-resolved
sudo rm /etc/resolv.conf
```

then create new /etc/resolv.conf with contents:
```
nameserver 8.8.8.8
nameserver 1.1.1.1
```

**Comment by alex207212** [comment]
I noticed that DNS will not work if "/run/systemd/resolve/resolv.conf" only has the following:
nameserver 192.168.0.1
search .
I then disconnect/connect AmneziaVPN client until I see this:
nameserver 192.168.0.1
nameserver 8.8.8.8
nameserver 8.8.4.4
search .
It looks like the client sometimes fails to update the DNS on local machine (8.8.*.* servers are the ones I have configured inside the client).
systemd-resolved is running on my machine

**Comment by alexeykuzmin** [comment]
On macOS make sure you're using a specific IP address in the client config's "Address" field, e.g. "10.0.0.2/32" (with 32 at the end), not a range like "10.0.0.2/24" (with 24 or anything else). 

It seems like the macOS app parses the "Address" contents as a subnet, then takes the network prefix (e.g. "10.0.0.0") and uses it as a client IP address. Which works well only if the subnet consists of a single address.
https://github.com/amnezia-vpn/amnezia-client/blob/f6e7d3ccf13afc23b1a4a73d650c03a187314d03/client/platforms/macos/daemon/iputilsmacos.cpp#L95-L97

**Comment by alex207212** [comment]
What is it all about?

**Comment by andreysum** [comment]
> `sudo apt install openvpn-systemd-resolved` did it for me (Ubuntu 24). Don't know what is DNS leak though :)

Works for me (Ubuntu 22)

**Comment by alex207212** [comment]
Here is what works for me:

1. After turning on Amnezia check `tail /run/systemd/resolve/resolv.conf` (in my case the nameservers are at the end of a file)
2. If you're missing DNS servers or see only 192.168.0.1 (maybe some other in your case like 192.168.xxx.xxx), then run the following script:
```
sudo systemctl restart systemd-resolved
sudo resolvectl dns wlp2s0 8.8.8.8 8.8.4.4
resolvectl status
```

Replace 8.8.8.8 8.8.4.4 with the servers you use and replace wlp2s0 with your interface name, run the following to see your interfaces:
`ip -br a`

**Comment by tyrypyrking** [comment]
What worked for me:
TL;DR: **Turn off killswitch**

With default configuration I see the following picture:
```
>nslookup google.com
;; communications error to 192.168.50.250#53: connection refused
;; communications error to 192.168.50.250#53: connection refused
;; communications error to 192.168.50.250#53: connection refused
;; no servers could be reached
```
`192.168.50.250` is my local DNS server that I have in `/etc/resolve.conf`. Turning the vpn does not update this record.

How to fix this: 
Cog icon (settings) -> Connection -> KillSwitch -> **Turn it off**

Try in now.
If you still wish to have both a killswitch and a local DNS then your best bet is setting custom DNS rules for local domains and removing `iptables` rule as suggested above to avoid this behavior.
